src/order/well_founded_set.lean

```diff
@@ -87,24 +87,45 @@ instance is_strict_order.subset {s : set α} {r : α → α → Prop} [is_strict
   to_is_trans := ⟨λ a b c ab bc, ⟨trans_of r ab.1 bc.1, ab.2.1, bc.2.2⟩ ⟩ }
 
 theorem well_founded_on_iff_no_descending_seq {s : set α} {r : α → α → Prop} [is_strict_order α r] :
-  s.well_founded_on r ↔ ∀ (f : ((>) : ℕ → ℕ → Prop) ↪r r), ¬ (range f) ⊆ s :=
+  s.well_founded_on r ↔ ∀ (f : ((>) : ℕ → ℕ → Prop) ↪r r), ¬∀ n, f n ∈ s :=
 begin
-  rw [well_founded_on_iff, rel_embedding.well_founded_iff_no_descending_seq],
-  refine ⟨λ h f con, begin
-      refine h.elim' ⟨⟨f, f.injective⟩, λ a b, _⟩,
-       simp only [con (mem_range_self a), con (mem_range_self b), and_true, gt_iff_lt,
-        function.embedding.coe_fn_mk, f.map_rel_iff]
-    end, λ h, ⟨λ con, _⟩⟩,
-  rcases con with ⟨f, hf⟩,
-  have hfs' : ∀ n : ℕ, f n ∈ s := λ n, (hf.2 n.lt_succ_self).2.2,
-  refine h ⟨f, λ a b, _⟩ (λ n hn, _),
-  { rw ← hf,
-    exact ⟨λ h, ⟨h, hfs' _, hfs' _⟩, λ h, h.1⟩ },
-  { rcases set.mem_range.1 hn with ⟨m, hm⟩,
-    rw ← hm,
-    apply hfs' }
+  simp only [well_founded_on_iff, rel_embedding.well_founded_iff_no_descending_seq, ← not_exists,
+    ← not_nonempty_iff, not_iff_not],
+  split,
+  { rintro ⟨⟨f, hf : ∀ {m n}, (r (f m) (f n) ∧ f m ∈ s ∧ f n ∈ s) ↔ n < m⟩⟩,
+    have H : ∀ n, f n ∈ s, from λ n, (hf.2 n.lt_succ_self).2.2,
+    refine ⟨⟨f, _⟩, H⟩,
+    simpa only [H, and_true] using @hf },
+  { rintro ⟨⟨f, hf⟩, hfs : ∀ n, f n ∈ s⟩,
+    refine ⟨⟨f, _⟩⟩,
+    simpa only [hfs, and_true] using @hf },
+end
+
+theorem well_founded_on.mono {s t : set α} {r r' : α → α → Prop} (h : t.well_founded_on r')
+  (hle : subrelation r r') (hst : s ⊆ t) : s.well_founded_on r :=
+begin
+  rw well_founded_on_iff at *,
+  refine subrelation.wf (λ x y xy, _) h,
+  exact ⟨hle xy.1, hst xy.2.1, hst xy.2.2⟩
 end
 
+theorem well_founded_on.mono_set {s t : set α} {r : α → α → Prop} (h : t.well_founded_on r)
+  (hst : s ⊆ t) : s.well_founded_on r :=
+h.mono (λ x y, id) hst
+
+theorem well_founded_on.union {s t : set α} {r : α → α → Prop} [is_strict_order α r]
```

Comment:
Maybe you could pull out `{s t : set α} {r : α → α → Prop}` to the top? It would unclutter the statements. And `β`, `ι` while you're at it.

Reply:
Reiterating, especially now that you made a section for `is_strict_order α r`.